torch/_inductor/codegen/cuda/__init__.py

```diff
@@ -0,0 +1,13 @@
+import os
+import sys
+
+THIRD_PARTY_PATH = "../../../../third_party/"
+CUTLASS_PY_PATH = "cutlass/tools/library/scripts/"
+
+# Add CUTLASS Python scripts to Python search path.
+sys.path.insert(
```

Comment:
A global modification of the sys.path is a very dangerous thing to do, and unexpected to happen on module load. Especially considering that namespace clashes are likely, considering what is below that path. See https://github.com/NVIDIA/cutlass/tree/main/tools/library/scripts - that location would override global packages named "library", "rt", "manifest", "generator" and even the top level __init__.py

While I see that there seems to be no easy workaround, I think this is too dangerous to merge into pytorch as is.

Reply:
Yes, this is a very valid concern.

1) I tried to update cutlass repo to use relative imports. However this breaks cutlass profiler so I had to revert this commit (https://github.com/NVIDIA/cutlass/compare/97b46e7549459f626f6b5c343acb6d75d089429d..12e5e4beb9dcc71af9708c80a5fb0168e599d555).
2) There are some discussions in the cutlass slack channel. Currently a preferred way is to add a new folder like "cutlass_scripts", create __init__.py under this folder and import submodules correctly.
3) It would take some time to make this change in cutlass. e.g. submit a PR, merge it in cutlass, sync it to Pytorch and internal, and update this piece of code. I don't want this PR to be blocked by the cutlass change.
4) I've also checked that under torch there are no existing libs naming like "library", "rt", "manifest", "generator". So it should be safe for now.
5) I'll work on the cutlass fix in parallel.

Reply:
Can we, maybe, (temporarily) get away with "private" imports? Something along the lines of importing the module with a helper function using `importlib` with the path to the module (possibly with `@cache`-ing). Would that work, without polluting the global `sys.modules` namespace?

Reply:
I'm not sure whether importlib works. I tried several tricks listed here but all failed: https://www.tutorialspoint.com/How-to-import-a-Python-module-given-the-full-path#:~:text=Using%20importlib.&text=of%20using%20importlib.-,util.,the%20module%20to%20be%20imported.
The reason is that these cutlass scripts import from other cutlass scripts, and during this process it still tries to find other cutlass scripts from path. It seems that the current way to import a module from a custom path doesn't work recursively.

Reply:
You might be able to do something similar to:
https://github.com/pytorch/pytorch/blob/1b2eac00cb6ebf7f51e8c7ed182b8915fa816f2b/test/inductor/test_compiled_autograd.py#L343-L348

It would be much better to put the cutlass stuff in a namespace rather than polluting the global one with these generic names.  There might be user code that uses these same names.

Reply:
Ah I see, I think `mock.patch` should work since it updates `sys.path`...
On the other hand I just submitted a PR to cutlass: https://github.com/NVIDIA/cutlass/pull/1072. If it could be accepted I'll just import from `third_party.cutlass.tools.library.scripts` directly.

Reply:
I see a potential problem with the approach of patching sys.path and using importlib's SourceFileLoader: If I get it correctly, using an importlib.SourceFileLoader loader will not bypass importlib's caches, nor the global sys.modules. Even worse, it will populate these. See https://peps.python.org/pep-0302/

"If there is an existing module object named ‘fullname’ in sys.modules, the loader must use that existing module. (Otherwise, the reload() builtin will not work correctly.) If a module named ‘fullname’ does not exist in sys.modules, the loader must create a new module object and add it to sys.modules."

The SourceFileLoader would therefore still contaminate the global module scope ( and also read from it ).

So, having a clean package name would be preferable. If we don't have that, we might have to resolve to some intermediate hack, like copying all source files into a new directory (say "cutlass_scripts"), and replacing every line like "import library" to "import cutlass_scripts.library as library" (etc..) within these files. Could be done with a few simple regex replacements as we can easily test that it works on the few files we would have to apply this to. It could also be done in a packaging script, not necessarily at runtime.

Reply:
This PR was updated last week to not use importlib's SourceFileLoader. It's now copying cutlass directory and adding cutlass_ prefixes to all submodules.